api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,383 @@
+	// Teleport Clients know to reuse the stored key instead of genearting a new one.
+	priv, err := yk.PrivateKey(slot, pub, piv.KeyAuth{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := selfSignedTeleportClientCertificate(priv, pub)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Store a self-signed certificate to mark this slot as used by tsh.
+	if err = yk.SetCertificate(piv.DefaultManagementKey, slot, cert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newYubiKeyPrivateKey(y, slot, pub)
+}
+
+// getPrivateKey gets an existing private key from the given PIV slot.
+func (y *yubiKey) getPrivateKey(slot piv.Slot) (*YubiKeyPrivateKey, error) {
+	yk, err := y.open()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	// Check the slot's certificate to see if it contains a self signed Teleport Client cert.
+	cert, err := yk.Certificate(slot)
+	if err != nil || cert == nil {
+		return nil, trace.NotFound("YubiKey certificate slot is empty, expected a Teleport Client cert")
+	} else if len(cert.Subject.Organization) == 0 || cert.Subject.Organization[0] != certOrgName {
+		return nil, trace.NotFound("YubiKey certificate slot contained unknown certificate:\n%+v", cert)
+	}
+
+	// Attest the key to make sure it hasn't been imported.
+	slotCert, err := yk.Attest(slot)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	attestationCert, err := yk.AttestationCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if _, err = piv.Verify(attestationCert, slotCert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Verify that the slot's certs have the same public key, otherwise the key
+	// may have been generated by a non-teleport client.
+	if pubComparer, ok := cert.PublicKey.(interface{ Equal(x crypto.PublicKey) bool }); !ok {
+		return nil, trace.BadParameter("certificate's public key of type %T is not a supported public key", cert.PublicKey)
+	} else if !pubComparer.Equal(slotCert.PublicKey) {
+		return nil, trace.NotFound("YubiKey slot contains mismatched certificates and must be regenerated")
+	}
+
+	return newYubiKeyPrivateKey(y, slot, slotCert.PublicKey)
+}
+
+// open a connection to yubiKey PIV module. The returned connection should be closed once
+// it's been used. The yubiKey PIV module itself takes some additional time to handle closed
+// connections, so we use a retry loop to give the PIV module time to close prior connections.
+func (y *yubiKey) open() (yk *piv.YubiKey, err error) {
+	isRetryError := func(err error) bool {
+		retryError := "connecting to smart card: the smart card cannot be accessed because of other connections outstanding"
+		return strings.Contains(err.Error(), retryError)
+	}
+
+	const maxRetries = 100
+	for i := 0; i < maxRetries; i++ {
+		yk, err = piv.Open(y.card)
+		if err == nil {
+			return yk, nil
+		}
+
+		if !isRetryError(err) {
+			return nil, trace.Wrap(err)
+		}
+
+		time.Sleep(time.Millisecond * 100)
```

Comment:
1. You're probably right, I'll move more Jitter/Retry logic from lib to api. Looks like some jitter logic was already moved.
2. Sure, good idea
3. We'll have to add ctx as an argument for all of the key parsing functions, but you're probably right that it's necessary. 

FYI: The yubikey tests require an actual Yubikey be attached, so `yubikey_test.go` won't be running in CI. Existing tests continue to use RSA private keys, since there isn't a yubikey PIV test module (that I could find).

Reply:
I went ahead and added in proper retry/context logic, but I decided not to pipe the context through all yubikey code for now to prevent over sizing this PR.

The connection logic is currently a work in progress, and I'm planning to create a follow up PR to manage open PIV connections instead of creating connections ad-hoc.